File: web/src/services/panier-local.ts
import { Injectable, computed, signal } from '@angular/core';

/** Une ligne telle que le navigateur la garde. */
export interface LigneLocale {
  readonly varianteId: number;
  readonly quantite: number;
  /** Recopiés pour AFFICHER le panier hors connexion — jamais pour calculer. */
  readonly nomProduit: string;
  readonly libelleDeclinaison: string;
  readonly urlPhoto: string | null;
  readonly prixIndicatif: number;
}

const CLE = 'garah.panier';
const MAX_LIGNES = 100;

/**
 * Le panier d'un visiteur non connecté.
 *
 * <h2>Pourquoi il existe</h2>
 *
 * <p>La vitrine est ouverte, le paiement non (D-07). Forcer la connexion au
 * premier « Ajouter » serait franc, mais ferait fuir le visiteur qui découvre
 * — et sur cet axe, le premier contact vient souvent d'un lien partagé, sans
 * compte.</p>
 *
 * <h2>⚠️ CE PANIER N'EST JAMAIS LA VÉRITÉ</h2>
 *
 * <p>Il ne connaît ni le stock, ni le prix courant, ni ce que le client a mis
 * de côté depuis un autre appareil. {@link prixIndicatif} sert à afficher un
 * total approximatif hors connexion, <b>jamais</b> à engager quoi que ce
 * soit : le serveur recalcule tout, et fige les montants au passage de
 * commande.</p>
 *
 * <p>À la connexion, {@code POST /api/panier/fusion} reprend ces lignes et
 * rend les <b>écarts</b> — article retiré, quantité déjà plus grande
 * ailleurs. C'est l'écran qui les annonce.</p>
 */
@Injectable({ providedIn: 'root' })
export class PanierLocal {
  private readonly lignesInternes = signal<readonly LigneLocale[]>(lire());

  readonly lignes = this.lignesInternes.asReadonly();

  readonly nombreArticles = computed(() =>
    this.lignesInternes().reduce((total, l) => total + l.quantite, 0),
  );

  /**
   * Le total <b>indicatif</b>.
   *
   * <p>⚠️ Il ignore les paliers de quantité et les frais d'acheminement : le
   * serveur, lui, applique le bon palier et ajoute les frais du point choisi.
   * L'écran doit donc le présenter comme une estimation, jamais comme un prix
   * à payer.</p>
   */
  readonly totalIndicatif = computed(() =>
    this.lignesInternes().reduce((total, l) => total + l.prixIndicatif * l.quantite, 0),
  );

  /**
   * Ajoute une quantité, ou crée la ligne.
   *
   * <p>Additif ici, contrairement à la fusion : le visiteur voit son panier,
   * il sait ce qu'il ajoute.</p>
   */
  ajouter(ligne: LigneLocale): void {
    this.lignesInternes.update((lignes) => {
      const existante = lignes.find((l) => l.varianteId === ligne.varianteId);

      if (existante) {
        return lignes.map((l) =>
          l.varianteId === ligne.varianteId ? { ...l, quantite: l.quantite + ligne.quantite } : l,
        );
      }

      // Le plafond protège d'un stockage local gonflé : le serveur refuse
      // au-delà de cent lignes, autant ne pas construire un panier qu'il
      // rejettera.
      if (lignes.length >= MAX_LIGNES) {
        return lignes;
      }
      return [...lignes, ligne];
    });
    this.enregistrer();
  }

  /** Fixe la quantité. Zéro retire la ligne — le geste attendu. */
  definirQuantite(varianteId: number, quantite: number): void {
    this.lignesInternes.update((lignes) =>
      quantite <= 0
        ? lignes.filter((l) => l.varianteId !== varianteId)
        : lignes.map((l) => (l.varianteId === varianteId ? { ...l, quantite } : l)),
    );
    this.enregistrer();
  }

  retirer(varianteId: number): void {
    this.lignesInternes.update((lignes) => lignes.filter((l) => l.varianteId !== varianteId));
    this.enregistrer();
  }

  /**
   * Vide le panier.
   *
   * <p>⚠️ Appelé après une fusion réussie <b>et</b> à la déconnexion. Le
   * second cas n'est pas du zèle : sur un poste partagé, le panier d'un
   * visiteur ne doit pas accueillir le suivant.</p>
   */
  vider(): void {
    this.lignesInternes.set([]);
    this.enregistrer();
  }

  /** Ce qu'on envoie à la fusion : le serveur n'a besoin que de deux champs. */
  pourFusion(): { varianteId: number; quantite: number }[] {
    return this.lignesInternes().map((l) => ({ varianteId: l.varianteId, quantite: l.quantite }));
  }

  private enregistrer(): void {
    try {
      localStorage.setItem(CLE, JSON.stringify(this.lignesInternes()));
    } catch {
      // Navigation privée, quota plein, stockage bloqué : le panier reste en
      // mémoire pour cette visite. Faire échouer l'ajout serait pire — le
      // visiteur ne comprendrait pas pourquoi le bouton ne fait rien.
    }
  }
}

/**
 * Relit le panier au démarrage.
 *
 * <p>⚠️ Tout ce qui vient de {@code localStorage} est <b>suspect</b> : une
 * version précédente de l'application, une main curieuse, un quota tronqué. On
 * valide chaque ligne plutôt que de faire confiance à la forme — une quantité
 * négative ou un identifiant absent produirait un panier que le serveur
 * refuserait, sans que l'écran sache pourquoi.</p>
 */
function lire(): readonly LigneLocale[] {
  try {
    const brut = localStorage.getItem(CLE);
    if (!brut) {
      return [];
    }

    const lu: unknown = JSON.parse(brut);
    if (!Array.isArray(lu)) {
      return [];
    }

    return lu.filter(estLigneValide).slice(0, MAX_LIGNES);
  } catch {
    return [];
  }
}

function estLigneValide(valeur: unknown): valeur is LigneLocale {
  if (typeof valeur !== 'object' || valeur === null) {
    return false;
  }
  const l = valeur as Record<string, unknown>;
  return (
    typeof l['varianteId'] === 'number' &&
    Number.isFinite(l['varianteId']) &&
    typeof l['quantite'] === 'number' &&
    Number.isInteger(l['quantite']) &&
    l['quantite'] > 0 &&
    typeof l['nomProduit'] === 'string'
  );
}

File: mobile/lib/services/panier_local.dart
import 'dart:convert';

import 'package:flutter/foundation.dart';
import 'package:shared_preferences/shared_preferences.dart';

/// Une ligne du panier local.
///
/// [prixIndicatif] est bien indicatif : le serveur recalcule au bon palier et
/// fige le prix à la commande. L'afficher comme un engagement mentirait le
/// jour où un tarif change entre l'ajout et le paiement.
class LigneLocale {
  const LigneLocale({
    required this.varianteId,
    required this.quantite,
    required this.nomProduit,
    required this.libelleDeclinaison,
    required this.urlPhoto,
    required this.prixIndicatif,
  });

  final int varianteId;
  final int quantite;
  final String nomProduit;
  final String? libelleDeclinaison;
  final String? urlPhoto;
  final num? prixIndicatif;

  LigneLocale avecQuantite(int q) => LigneLocale(
    varianteId: varianteId,
    quantite: q,
    nomProduit: nomProduit,
    libelleDeclinaison: libelleDeclinaison,
    urlPhoto: urlPhoto,
    prixIndicatif: prixIndicatif,
  );

  Map<String, dynamic> versJson() => {
    'varianteId': varianteId,
    'quantite': quantite,
    'nomProduit': nomProduit,
    'libelleDeclinaison': libelleDeclinaison,
    'urlPhoto': urlPhoto,
    'prixIndicatif': prixIndicatif,
  };

  factory LigneLocale.de(Map<String, dynamic> j) => LigneLocale(
    varianteId: (j['varianteId'] as num).toInt(),
    quantite: (j['quantite'] as num).toInt(),
    nomProduit: (j['nomProduit'] as String?) ?? 'Article',
    libelleDeclinaison: j['libelleDeclinaison'] as String?,
    urlPhoto: j['urlPhoto'] as String?,
    prixIndicatif: j['prixIndicatif'] as num?,
  );
}

/// Le panier qui vit dans le téléphone.
///
/// ## 🎯 On remplit son panier SANS COMPTE
///
/// Exiger la connexion pour ajouter un article ferait fuir au premier
/// « Ajouter ». La connexion arrive au moment de commander, pas avant — et à
/// ce moment-là le panier local est FUSIONNÉ avec celui du serveur.
///
/// ## ⚠️ Ce panier n'est JAMAIS la vérité
///
/// Le serveur décide du stock et du prix. Lui faire confiance, c'est vendre ce
/// qu'on n'a plus. Il ne sert qu'à ne pas perdre ce qu'on a choisi.
class PanierLocal extends ChangeNotifier {
  static const _cle = 'garah.panier';

  final List<LigneLocale> _lignes = [];

  List<LigneLocale> get lignes => List.unmodifiable(_lignes);

  int get nombreArticles => _lignes.fold(0, (somme, l) => somme + l.quantite);

  num get sousTotal => _lignes.fold<num>(
    0,
    (somme, l) => somme + (l.prixIndicatif ?? 0) * l.quantite,
  );

  Future<void> relire() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final brut = prefs.getString(_cle);
      if (brut == null) return;
      _lignes
        ..clear()
        ..addAll(
          (jsonDecode(brut) as List<dynamic>).map(
            (e) => LigneLocale.de(e as Map<String, dynamic>),
          ),
        );
      notifyListeners();
    } catch (_) {
      // Stockage illisible ou format d'une version précédente : on repart d'un
      // panier vide plutôt que d'empêcher l'application de démarrer.
      _lignes.clear();
    }
  }

  /// Ajoute, ou RELÈVE la quantité si l'article y est déjà.
  ///
  /// ⚠️ On ne cumule pas les deux quantités : sur une connexion instable, le
  ///    même ajout part parfois deux fois. Prendre la plus grande rend le
  ///    geste idempotent ; additionner ferait grimper le panier tout seul.
  Future<void> ajouter(LigneLocale ligne) async {
    final i = _lignes.indexWhere((l) => l.varianteId == ligne.varianteId);
    if (i >= 0) {
      final existante = _lignes[i];
      if (ligne.quantite > existante.quantite) {
        _lignes[i] = existante.avecQuantite(ligne.quantite);
      }
    } else {
      _lignes.add(ligne);
    }
    await _ecrire();
  }

  Future<void> changerQuantite(int varianteId, int quantite) async {
    final i = _lignes.indexWhere((l) => l.varianteId == varianteId);
    if (i < 0) return;
    if (quantite <= 0) {
      _lignes.removeAt(i);
    } else {
      _lignes[i] = _lignes[i].avecQuantite(quantite);
    }
    await _ecrire();
  }

  Future<void> retirer(int varianteId) async {
    _lignes.removeWhere((l) => l.varianteId == varianteId);
    await _ecrire();
  }

  Future<void> vider() async {
    _lignes.clear();
    await _ecrire();
  }

  /// Ce qu'on envoie au serveur à la connexion.
  List<Map<String, dynamic>> pourFusion() => _lignes
      .map((l) => {'varianteId': l.varianteId, 'quantite': l.quantite})
      .toList();

  Future<void> _ecrire() async {
    notifyListeners();
    try {
      final prefs = await SharedPreferences.getInstance();
      await prefs.setString(
        _cle,
        jsonEncode(_lignes.map((l) => l.versJson()).toList()),
      );
    } catch (_) {
      // Écriture refusée : le panier vaut pour cette session. Faire échouer
      // l'ajout serait pire — le bouton ne ferait rien.
    }
  }
}
